kubetest2-gke/deployer/up.go

```diff
@@ -53,6 +66,53 @@ func (d *deployer) Up() error {
 	if err := d.prepareGcpIfNeeded(d.projects[0]); err != nil {
 		return err
 	}
+
+	var err error
+	if d.region != "" {
+		for i, r := range append([]string{d.region}, d.backupRegions...) {
+			if i != 0 {
+				klog.V(2).Infof("Retrying creating the clusters in the backup region %q", r)
+			}
+			d.region = r
+
+			if err = d.createClusters(); err == nil || !isRetryableError(err) {
+				break
+			}
+		}
+	} else {
+		for i, z := range append([]string{d.zone}, d.backupZones...) {
+			if i != 0 {
+				klog.V(2).Infof("Retrying creating the clusters in the backup zone %q", z)
+			}
+			d.zone = z
+
+			if err = d.createClusters(); err == nil || !isRetryableError(err) {
```

Comment:
this messes "down". we would need to delete resources from failed attempts as well.

Reply:
I don't have a good solution to this comment, but IMO we will only need `--backup-regions` and `--backup-zones` in CI since the flakiness is not really important if the users want to run the test flows on their local. And in CI we always have Boskos janitor to clean them up so it seems not a big issue.

Also the `down` flow is already broken since now it will throw an error when deleting clusters, if `up` failed to create the clusters due to quota issues, in which the networking and firewall rules will be leaked.

Reply:
Or we can add an extra check to disallow setting `--backup-regions` and `--backup-zones` when `CI` is not true - https://github.com/kubernetes/test-infra/blob/master/prow/jobs.md#job-environment-variables?

Reply:
Also if the two backup flags are not set, which is the default, the changes in this PR will just be a no-op. So it's not a breaking Chang for existing test flows.

Reply:
relying on janitor is not an ideal solution. janitor is only supposed to be used when say jobs are aborted. A successful kubetest2 --up --down should not rely on boskos janitor existing. 

Why is the down flow broken? if you mean that cluster delete command fails if cluster wasn't created in the first place then that's WAI?
We definitely don't take care of resources that are brought up behind the scenes during cluster creation, but if we are creating any resources/firewall rules as part of `Up` then those should be cleaned up in `Down` otherwise that's a bug.

currently as this stands if we have 2 clusters and 2 zones (1 original and 1 backup)
and cluster 1 is created in zone1 but cluster2 fails
we are retrying everything in zone2
and if both clusters come up in zone2 then cluster1 in zone1 is leaked.

Reply:
Currently for all the jobs we run with GKE clusters, we are not using the `--down` flag in order to save time in CI, so we are completely relying on janitor to do the cleanups, and we haven't seen any issues so far.

> Why is the down flow broken? if you mean that cluster delete command fails if cluster wasn't created in the first place then that's WAI?
Yes, the networks and firewall rules will be leaked if the cluster wasn't created in the first place, but we have janitor to clean them up so it's not an issue in CI.

It will complicate the logic a lot if we want to support cleaning up these leftover resources due to retrying, and also as I mentioned in another thread, this will be a no-op change if `--backup-regions` and `--backup-zones` are empty, which is by default, so it won't impact users who do not need it.

Maybe I can add description to these two flags saying `setting them will probably leak resources, please use them with caution`, and we can add a `TODO` to fix it? Or if you are completely not convinced with this change, I'll then update ASM/Istio test-infra, then we'll have this logic duplicated everywhere...

Reply:
I'm definitely +1 on having a retry mode, but would like to see the suggested changes :) any new feature should be self contained and honor the full kubetest2 cluster lifecycle. We shouldn't be adding features which explicitly rely on also needing janitor.

Is there any reason why all clusters need to be recreated in a new zone, can't we just recreate the ones that fail? 
also having backup zones tells me that your tests don't really have any preference of a zone. 
Then, we can also make --zone itself take a comma separated list and choose randomly among them (and then use retries in different zones)

I believe for non stockout retriable errors, users would also be interested in retrying cluster creation even for original single zone requests.

tldr; I would love to see this being added as a more general retry feature (with all these considerations) that isn't solely for getting around stockout errors. (stockout errors can still be one of the retriable errors)

Reply:
> Is there any reason why all clusters need to be recreated in a new zone, can't we just recreate the ones that fail?

I can try that, but looks it'll make the [Up](https://github.com/kubernetes-sigs/kubetest2/blob/master/kubetest2-gke/deployer/up.go#L37) function even more complicated, and also complicate the logic if we want to do proper cleanups since we'll need to track which regions/zones it tried to create for each cluster.

> also having backup zones tells me that your tests don't really have any preference of a zone.
Then, we can also make --zone itself take a comma separated list and choose randomly among them (and then use retries in different zones)

Currently for multicluster, we are assuming they should be created in the same region/zone, but in the future we might want them to be created in different regions/zones (the requirement was brought up by one of the engineer partners but it's not a high priority), then we'll need to change `--region` and `--zone` to take a comma separated list with its length equal to the number of clusters. That's why I'm not reusing this flag to set the backup regions/zones.

> I believe for non stockout retriable errors, users would also be interested in retrying cluster creation even for original single zone requests.

Other errors are mostly due to "the cluster was created, but the health check failed", so we will always get an error if we retry creating the cluster with **the same name** in the same region/zone. Then it'll require us deleting the clusters before recreating them, which is less efficient than directly recreating them in a different region/zone.


I can get your concern and I agree they are all valid, but I'm not sure what'll the best way to move forward without complicating the code logic here too much.

Reply:
Yes, I would like to see the retry feature be discussed in more detail (especially for multi cluster) before we try to add anything, maybe as an issue/doc to cover existing+future use cases (like you mentioned). 

We can think about reworking the flags to be read from a config file if we want a lot of customizability and have each cluster's lifecycle be handled independently.

Reply:
I've created https://github.com/kubernetes-sigs/kubetest2/issues/88, please take a look when you have time. Thanks.